src/opendlv-device-kiwi-prugw.cpp

```diff
@@ -15,25 +15,96 @@
+    std::cerr << "Usage:   " << argv[0] << " --cid=<OpenDaVINCI session> --names=<Strings> --types=<esc or servo> --channels=<1...8>  --offsets<-1...1> --maxvals=<floats> --angleconversion=<const> [--verbose]" << std::endl;
+    std::cerr << "Example: " << argv[0] << " --cid=111 --names=steering,propulsion --types=servo,esc --channels=1,2 --offsets=0,0 --maxvals=0.5,0 --angleconversion=1" << std::endl;
     retCode = 1;
   } else {
+
+    // Setup
+    int32_t VERBOSE{commandlineArguments.count("verbose") != 0};
+    if (VERBOSE) {
+      VERBOSE = std::stoi(commandlineArguments["verbose"]);
+    }
+
+    std::vector<std::string> names = stringtoolbox::split(commandlineArguments["names"],',');
+    std::vector<std::string> types = stringtoolbox::split(commandlineArguments["types"],',');
+    std::vector<std::string> channels = stringtoolbox::split(commandlineArguments["channels"],',');
+    std::vector<std::string> offets = stringtoolbox::split(commandlineArguments["offsets"],',');
+    std::vector<std::string> maxvals = stringtoolbox::split(commandlineArguments["maxvals"],',');
+    float const angleConversion = std::stof(commandlineArguments["angleconversion"]);
+
+    if (names.size() != types.size() ||
+        names.size() != types.size() ||
+        names.size() != channels.size() ||
+        names.size() != offets.size() ||
+        names.size() != maxvals.size()) {
+      std::cerr << "Number of arguments do not match, use ',' as delimiter." << std::endl;
+      retCode = 1;
+    }
+
+    PwmMotors pwmMotors(names, types, channels, offets, maxvals);
+    pwmMotors.powerServoRail(true);
+
     cluon::OD4Session od4{static_cast<uint16_t>(std::stoi(commandlineArguments["cid"])),
-      [](auto){}
+      [&pwmMotors, &angleConversion](cluon::data::Envelope &&envelope){
+        if (envelope.dataType() == opendlv::proxy::GroundSteeringRequest::ID()) {
+          opendlv::proxy::GroundSteeringRequest gst = cluon::extractMessage<opendlv::proxy::GroundSteeringRequest>(std::move(envelope));
+          float groundSteering = gst.groundSteering() / angleConversion;
+          pwmMotors.setMotorPower(1, groundSteering);
+        } else if (envelope.dataType() == opendlv::proxy::PedalPositionRequest::ID()) {
+          opendlv::proxy::PedalPositionRequest ppr = cluon::extractMessage<opendlv::proxy::PedalPositionRequest>(std::move(envelope));
+          float val = (ppr.position()+1)/2.0f;
+          if (val > 1.0f) {
+            val = 1.0f;
+          } else if (val < 0.1f){
+            val = 0.1f;
+          }
+          pwmMotors.setMotorPower(2, val);
+        }
+      }
     };
 
+    if (VERBOSE == 2) {
+      initscr();
+    }
+    // getch();       /Wait for user input 
+
     while (od4.isRunning()) {
-      std::this_thread::sleep_for(std::chrono::duration<double>(1.0));
+      // This must be called regularly (>40hz) to keep servos or ESCs awake.
+      std::this_thread::sleep_for(std::chrono::duration<double>(1.0 / 50.0f));
```

Comment:
This freq computation does not compensate for the time spent for the actual job. You need to subtract the time spent between two consecutive calls to sleep_for